File: sem4/mpp/prac1/server/src/main/java/server/UserService.java
package server;

import domain.*;
import domain.dto.ChildDTO;
import repository.ChildRepository;
import repository.SignupRepository;

import java.util.List;

public class UserService {
	private final ChildRepository childRepository;
	private final SignupRepository signupRepository;

	private static final int MAX_SIGNUPS = 2;

	public UserService(ChildRepository childRepository, SignupRepository signupRepository) {
		this.childRepository = childRepository;
		this.signupRepository = signupRepository;
	}

	public synchronized List<ChildDTO> searchChildren(RaceDistance raceDistance, AgeGroup ageGroup) {
		return signupRepository.getChildrenByRace(ageGroup, raceDistance)
				.stream()
				.map(c -> new ChildDTO(c.getName(), c.getAge(), signupRepository.getChildSignups(c.getId())))
				.toList();
	}

	public synchronized void signUpChild(String name, Integer age, RaceDistance raceDistance, AgeGroup ageGroup) {
		if (ageGroup == null)
			throw new RuntimeException("Invalid age group");
		Child child = childRepository.getChildByNameAge(name, age);
		if (child == null)
			child = childRepository.add(new Child(name, age));
		if (signupRepository.getChildSignups(child.getId()) >= MAX_SIGNUPS)
			throw new RuntimeException("Child has signed up to the maximum of " + MAX_SIGNUPS + " races already.");
		signupRepository.add(new Signup(child, ageGroup, raceDistance));
	}
}

File: sem4/mpp/project-cs/server/service/UserService.cs
using project_cs.domain;
using project_cs.repo;

namespace project_cs.service;

public class UserService {
    private readonly IChildRepository _childRepository;
    private readonly ISignupRepository _signupRepository;

    private const int MaxSignups = 2;

    public UserService(IChildRepository childRepository, ISignupRepository signupRepository) {
        this._childRepository = childRepository;
        this._signupRepository = signupRepository;
    }

    public List<ChildDto> SearchChildren(RaceDistance raceDistance, AgeGroup ageGroup) {
        return _signupRepository.GetChildrenByRace(ageGroup, raceDistance)
            .Select(c => new ChildDto(c.name, c.age, _signupRepository.GetChildSignups(c.id)))
            .ToList<ChildDto>();
    }

    public void SignUpChild(string name, int age, RaceDistance raceDistance, AgeGroup? ageGroup) {
        if (ageGroup == null)
            throw new Exception("Invalid age group");
        var child = _childRepository.GetChildByNameAge(name, age);
        if (child == null)
            child = _childRepository.Add(new Child(name, age));
        if (_signupRepository.GetChildSignups(child.id) >= MaxSignups)
            throw new Exception("Child has signed up to the maximum of " + MaxSignups + " races already.");
        _signupRepository.Add(new Signup(child, (AgeGroup)ageGroup, raceDistance));
    }
}